src/decoder/mp3.rs

```diff
@@ -1,129 +1,917 @@
+    /// When this reaches the frame's sample count, a new frame must be decoded.
     current_span_offset: usize,
+
+    /// Number of audio channels.
+    ///
+    /// Can theoretically change between frames, though this is rare in practice.
+    /// Updated dynamically when frame channel count changes.
+    channels: ChannelCount,
+
+    /// Sample rate in Hz.
+    ///
+    /// Fixed for the entire MP3 stream. Common rates include 44.1kHz (CD quality),
+    /// 48kHz (professional), and various rates for different MPEG versions.
+    sample_rate: SampleRate,
+
+    /// Number of samples read so far (for seeking calculations).
+    ///
+    /// Tracks the current playback position in total samples (across all channels).
+    /// Used to determine if seeking requires stream reset or can skip forward.
+    samples_read: u64,
+
+    /// Total number of audio samples (estimated from duration).
+    ///
+    /// Calculated from total duration when available. For VBR files without
+    /// metadata, this may be an estimate based on average bitrate calculations.
+    total_samples: Option<u64>,
+
+    /// Total duration calculated from file analysis or metadata.
+    ///
+    /// Can be provided explicitly, calculated via duration scanning, or estimated
+    /// from file size and average bitrate. Most accurate when obtained through
+    /// full file scanning.
+    total_duration: Option<Duration>,
+
+    /// Average bitrate in bytes per second (estimated).
+    ///
+    /// Updated dynamically during playback to improve seeking accuracy.
+    /// Initial value comes from first frame or duration/size calculation.
+    average_bitrate: u32,
+
+    /// MPEG layer (typically 3 for MP3).
+    ///
+    /// Determines frame structure and sample count per frame:
+    /// - Layer I: 384 samples per frame
+    /// - Layer II/III: 1152 samples per frame
+    mpeg_layer: usize,
+
+    /// Seeking precision mode.
+    ///
+    /// Controls the trade-off between seeking speed and accuracy:
+    /// - `Fastest`: Byte-position-based seeking (fastest but potentially inaccurate for VBR)
+    /// - `Nearest`: Sample-accurate seeking (slower but always accurate)
+    seek_mode: SeekMode,
+
+    /// Whether random access seeking is enabled.
+    ///
+    /// When `true`, enables backward seeking by allowing stream reset operations.
+    /// When `false`, only forward seeking (sample skipping) is allowed.
+    is_seekable: bool,
 }
 
 impl<R> Mp3Decoder<R>
 where
     R: Read + Seek,
 {
-    pub fn new(mut data: R) -> Result<Self, R> {
-        if !is_mp3(data.by_ref()) {
+    /// Attempts to decode the data as MP3 with default settings.
+    ///
+    /// This method probes the input data to detect MP3 format and initializes the decoder if
+    /// successful. Uses default settings with no seeking support or duration scanning enabled.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - Input stream implementing `Read + Seek`
+    ///
+    /// # Returns
```

Comment:
Again the java style doc comments. This is a very very verbose way of saying what the function signature is